File: benchmark/python/lib/benchmark_runner.py
"""
Core benchmark execution framework for Python/NumPy.
"""

import time
from typing import Callable, TypeVar, Optional, Any
from dataclasses import dataclass

T = TypeVar('T')


@dataclass
class BenchmarkResult:
    """Result from a single benchmark run."""
    mean_ms: float
    min_ms: float
    max_ms: float
    std_ms: float
    iterations: int
    result: Optional[float] = None


class BenchmarkRunner:
    """
    Benchmark runner with warmup, timing, and statistics calculation.
    """

    def __init__(self, config: dict):
        self.warmup_iterations = config.get('warmup_iterations', 3)
        self.min_iterations = config.get('min_iterations', 10)
        self.min_duration_sec = config.get('min_duration_sec', 1.0)

    def run(
        self,
        bench_fn: Callable[[], Any],
        cleanup_fn: Optional[Callable[[], None]] = None
    ) -> BenchmarkResult:
        """
        Run a benchmark function.

        Args:
            bench_fn: Function to benchmark (called repeatedly)
            cleanup_fn: Optional cleanup function (called once after benchmarking)

        Returns:
            BenchmarkResult with timing statistics
        """
        # Warmup runs (not counted)
        for _ in range(self.warmup_iterations):
            bench_fn()

        # Timed runs - adaptive iteration count
        times = []
        start_total = time.perf_counter()
        result = None

        while (len(times) < self.min_iterations or
               time.perf_counter() - start_total < self.min_duration_sec):
            start = time.perf_counter()
            result = bench_fn()
            end = time.perf_counter()
            times.append((end - start) * 1000)  # Convert to milliseconds

        # Cleanup
        if cleanup_fn:
            cleanup_fn()

        return self._compute_stats(times, result)

    def _compute_stats(self, times: list[float], result: Any) -> BenchmarkResult:
        """Compute statistics from timing data."""
        n = len(times)
        mean_ms = sum(times) / n
        variance = sum((t - mean_ms) ** 2 for t in times) / n

        # Convert result to float if it's a scalar
        result_val = None
        if result is not None:
            try:
                result_val = float(result)
            except (TypeError, ValueError):
                pass

        return BenchmarkResult(
            mean_ms=mean_ms,
            min_ms=min(times),
            max_ms=max(times),
            std_ms=variance ** 0.5,
            iterations=n,
            result=result_val,
        )


def create_runner(config: Optional[dict] = None) -> BenchmarkRunner:
    """Create a benchmark runner with default or custom configuration."""
    default_config = {
        'warmup_iterations': 3,
        'min_iterations': 10,
        'min_duration_sec': 1.0,
    }
    if config:
        default_config.update(config)
    return BenchmarkRunner(default_config)

File: benchmark/ts/lib/BenchmarkRunner.ts
/**
 * Core benchmark execution framework.
 * Handles warmup, timing, and statistics calculation.
 */

import { performance } from 'perf_hooks';
import type { BenchmarkConfig, BenchmarkResult } from './types.js';

export class BenchmarkRunner {
  private config: BenchmarkConfig;

  constructor(config: BenchmarkConfig) {
    this.config = config;
  }

  /**
   * Run a benchmark with setup, execution, and optional cleanup.
   *
   * @param setupFn - Function to prepare test data (not timed)
   * @param benchFn - Function to benchmark (timed)
   * @param cleanupFn - Optional cleanup function (not timed)
   */
  async run<T>(
    setupFn: () => T | Promise<T>,
    benchFn: (setup: T) => number | void,
    cleanupFn?: (setup: T) => void
  ): Promise<BenchmarkResult> {
    const setup = await setupFn();

    // Warmup runs (not counted)
    for (let i = 0; i < this.config.warmupIterations; i++) {
      benchFn(setup);
    }

    // Timed runs - adaptive iteration count
    const times: number[] = [];
    const startTotal = performance.now();
    let result: number | void = undefined;

    while (
      times.length < this.config.minIterations ||
      performance.now() - startTotal < this.config.minDurationMs
    ) {
      const start = performance.now();
      result = benchFn(setup);
      times.push(performance.now() - start);
    }

    // Cleanup
    if (cleanupFn) {
      cleanupFn(setup);
    }

    return this.computeStats(times, result!);
  }

  /**
   * Run a synchronous benchmark (simpler API for sync functions).
   */
  runSync<T>(
    setupFn: () => T,
    benchFn: (setup: T) => number | void,
    cleanupFn?: (setup: T) => void
  ): BenchmarkResult {
    const setup = setupFn();

    // Warmup runs (not counted)
    for (let i = 0; i < this.config.warmupIterations; i++) {
      benchFn(setup);
    }

    // Timed runs - adaptive iteration count
    const times: number[] = [];
    const startTotal = performance.now();
    let result: number | void = undefined;

    while (
      times.length < this.config.minIterations ||
      performance.now() - startTotal < this.config.minDurationMs
    ) {
      const start = performance.now();
      result = benchFn(setup);
      times.push(performance.now() - start);
    }

    // Cleanup
    if (cleanupFn) {
      cleanupFn(setup);
    }

    return this.computeStats(times, result!);
  }

  /**
   * Run an async benchmark where the benchmark function returns a Promise.
   * Used for operations that are inherently async (like WASM array creation).
   */
  async runAsync<T>(
    setupFn: () => T | Promise<T>,
    benchFn: (setup: T) => Promise<number | void>,
    cleanupFn?: (setup: T) => void | Promise<void>
  ): Promise<BenchmarkResult> {
    const setup = await setupFn();

    // Warmup runs (not counted)
    for (let i = 0; i < this.config.warmupIterations; i++) {
      await benchFn(setup);
    }

    // Timed runs - adaptive iteration count
    const times: number[] = [];
    const startTotal = performance.now();
    let result: number | void = undefined;

    while (
      times.length < this.config.minIterations ||
      performance.now() - startTotal < this.config.minDurationMs
    ) {
      const start = performance.now();
      result = await benchFn(setup);
      times.push(performance.now() - start);
    }

    // Cleanup
    if (cleanupFn) {
      await cleanupFn(setup);
    }

    return this.computeStats(times, result!);
  }

  private computeStats(times: number[], result?: number | void): BenchmarkResult {
    const n = times.length;
    const meanMs = times.reduce((a, b) => a + b, 0) / n;
    const variance = times.reduce((sum, t) => sum + (t - meanMs) ** 2, 0) / n;

    return {
      meanMs,
      minMs: times.reduce((a, b) => Math.min(a, b), Infinity),
      maxMs: times.reduce((a, b) => Math.max(a, b), -Infinity),
      stdMs: Math.sqrt(variance),
      iterations: n,
      result: typeof result === 'number' ? result : undefined,
    };
  }
}

/**
 * Create a default benchmark runner with standard configuration.
 */
export function createRunner(config?: Partial<BenchmarkConfig>): BenchmarkRunner {
  const defaultConfig: BenchmarkConfig = {
    warmupIterations: 3,
    minIterations: 10,
    minDurationMs: 1000,
  };

  return new BenchmarkRunner({ ...defaultConfig, ...config });
}
